packages/slate-react/src/utils/dom.ts

```diff
@@ -98,6 +98,15 @@ export const normalizeDOMPoint = (domPoint: DOMPoint): DOMPoint => {
   return [node, offset]
 }
 
+/**
+ * Returns ShadowRoot if active element is renderd within Shadow DOM,
+ * otherwise it returns the Document Object.
+ */
+
+export const getDocumentOrShadowRoot = (): Document | ShadowRoot => {
+  return window.document.activeElement?.shadowRoot ?? window.document
```

Comment:
This seems work for one level of shadowRoot but not if they are nested. I think it needs to be done bottom up from the element in question to find the shadowRoot that contains that element.

I've only come across searches up the node tree to find the appropriate root with something like below (illustrative only). It's probably something that's be computed once and stored on the editor for use when it's needed.

```
let n = editor;
while (n.
for (let n = n.parentNode; n; n = n.parentNode) {
  if (n instanceof Document || n instanceof ShadowRoot) break;
}
const documentOrShadowRoot = n || window.document;
```

Reply:
Hmm, interesting. Could you provide a simple example codesandbox (or something similar)?
You can look at the example I created in this PR as a starting point.

Reply:
I [forked your branch](https://github.com/AndrewScull/slate/commits/use_shadow_dom_if_available) of this pull request and updated the Shadow DOM example to also include a nested Shadow DOM.

Reply:
I've updated that branch with a commit that gets the nested version working too. It's far from being PR-ready and there are some behaviours that I can't yet fully explain so if you're more familiar with these APIs it'd be great to hear what you think.

Reply:
I had another go at this ([branch](https://github.com/AndrewScull/slate/tree/shadowdom)) after coming across `getRootNode()` to avoid the tree traversal. However, I'm still running into an issue with key events that seems to be a [React issue](https://github.com/facebook/react/issues/9242) with shadow DOM. On Chrome, the undo key combo doesn't trigger the event handler and so the undo doesn't happen.

Reply:
Nice work. I merged your changes into this PR. I've also updated a few smaller things (sorry for the mess with all those commits. I was trying out a few things...)